js/background.js

```diff
@@ -483,84 +499,93 @@
-        var type, id;
-        if (data.message.group) {
-            type = 'group';
-            id = data.message.group.id;
-        } else {
-            type = 'private';
-            id = data.destination;
-        }
-
-        if (data.message.flags & textsecure.protobuf.DataMessage.Flags.PROFILE_KEY_UPDATE) {
-            return ConversationController.getOrCreateAndWait(id, type).then(function(convo) {
-              return convo.save({profileSharing: true}).then(ev.confirm);
-            });
-        }
-
-        var message = new Whisper.Message({
-            source         : textsecure.storage.user.getNumber(),
-            sourceDevice   : data.device,
-            sent_at        : data.timestamp,
-            received_at    : now,
-            conversationId : data.destination,
-            type           : 'outgoing',
-            sent           : true,
-            expirationStartTimestamp: data.expirationStartTimestamp,
-        });
+  /* eslint-enable */
+  /* jshint ignore:start */
+  function createMessageHandler(
+    { createMessage, getMessageDescriptor, handleProfileUpdate }
+  ) {
+    return async (event) => {
+      const { data, confirm } = event;
+
+      const messageDescriptor = getMessageDescriptor(data);
+
+      const isProfileUpdate = Boolean(
+        // eslint-disable-next-line no-bitwise
+        data.message.flags & textsecure.protobuf.DataMessage.Flags.PROFILE_KEY_UPDATE
+      );
+      if (isProfileUpdate) {
+        return handleProfileUpdate({ data, confirm, messageDescriptor });
+      }
 
-        return isMessageDuplicate(message).then(function(isDuplicate) {
-            if (isDuplicate) {
-                console.log('Received duplicate message', message.idForLogging());
-                ev.confirm();
-                return;
-            }
+      const message = createMessage(data);
+      const isDuplicate = await isMessageDuplicate(message);
+      if (isDuplicate) {
+        console.log('Received duplicate message', message.idForLogging());
+        return event.confirm();
+      }
 
-            return ConversationController.getOrCreateAndWait(id, type).then(function() {
-                return message.handleDataMessage(data.message, ev.confirm, {
-                    initialLoadComplete: initialLoadComplete
-                });
-            });
-        });
-    }
+      const upgradedMessage = await Message.upgradeSchema(data.message);
+      await ConversationController.getOrCreateAndWait(
+        messageDescriptor.id, messageDescriptor.type
+      );
+      return message.handleDataMessage(
+        upgradedMessage,
+        event.confirm,
+        { initialLoadComplete }
+      );
+    };
+  }
+
+  // Received:
+  async function handleMessageReceivedProfileUpdate(
+    { data, confirm, messageDescriptor }
+  ) {
+    const profileKey = data.message.profileKey.toArrayBuffer();
+    const sender = await ConversationController.getOrCreateAndWait(
+      messageDescriptor.source, 'private'
```

Comment:
Is this `.source` a bug? My understanding is that message descriptors are always just a `type` and an `id`. At a higher level, I wonder if the term 'descriptor' needs a better definition somewhere, or if we could come up with a better term? In the same vein, the distance between the definition of the descriptors and their use here made it hard to find this bug. Maybe you could talk a little bit about your vision for descriptors?